src/coordinates.rs

```diff
@@ -0,0 +1,206 @@
+//! Extension traits and structs that provide generic access to the coordinates of
+//! elliptic curve points.
+//!
+//! Coordinates are meaningless without the context of the curve equation that constrains
+//! them. To safely expose them in a generic context, we use extension traits to restrict
+//! the scope of the generic curve parameter; this ensures that the code can only be used
+//! with curve implementations that explicitly expose their use of a specific curve model.
+
+use subtle::{Choice, ConditionallySelectable, CtOption};
+
+use crate::CurveAffine;
+
+//
+// Twisted Edwards curve
+//
+
+/// An affine elliptic curve point on a twisted Edwards curve
+/// $a \cdot x^2 + y^2 = 1 + d \cdot x^2 \cdot y^2$.
+pub trait TwistedEdwardsPoint: CurveAffine + Default + ConditionallySelectable {
+    /// Field element type used in the curve equation.
+    type Base: Copy + ConditionallySelectable;
+
+    /// The parameter $a$ in the twisted Edwards curve equation.
+    ///
+    /// When $a = 1$, this reduces to an ordinary Edwards curve.
+    const A: Self::Base;
+
+    /// The parameter $d$ in the twisted Edwards curve equation.
+    const D: Self::Base;
```

Comment:
I'm not yet sure whether this trait should include the model parts as well, or whether that should be via an extension trait of the `Curve` trait. In general downstream users should use the `Curve` trait as the primary downstream interface, which would mean applying a bound of the form:
```rust
fn foo<C: Curve>() where C::Affine: TwistedEdwardsPoint {}
```

However, if we're not going to expose model-related APIs on the `Curve` type, then this is probably overall a simpler trait hierarchy, and a reasonable pathway to take (that we could always change track on later after we get some experience with it).

Reply:
Question: is there a specific reason that `MontgomeryPoint` is necessarily an `Affine` subtype? It also has a projective representation. Would it make sense to separate the notion of curve equation and coordinate representation?

From an API perspective, I think that'd mean writing `trait MontgomeryPoint` as below, but without the methods. And then defining `struct AffineMontgomeryPoint<Base> { u: Base, v: Base }` and `impl<F> MontgomeryPoint<Base=F> for AffineMontgomeryPoint<F>` or something like that. Does that make sense?

Reply:
> Question: is there a specific reason that `MontgomeryPoint` is necessarily an `Affine` subtype? It also has a projective representation. Would it make sense to separate the notion of curve equation and coordinate representation?

This is somewhat an artifact of the trait approach we've taken in the `group` crate: the `Group` trait denotes both a cryptographic group, and elements of that group. So there is currently no single place where a curve-only trait could live.

What I was suggesting above was that we would effectively do this split, by way of having an extension trait of `Curve` where the curve model lives, but that trait would also be where any efficient-representation methods would need to live. Except that the efficient representation is _not_ explicitly any specific representation; it's "whatever representation is best for this curve implementation", meaning that there's nothing to expose there.

So if we wanted to expose the projective representation, it would likely instead be an extension trait to the extension trait, which starts making the trait hierarchy quite complex, and we'd need a way to convey and manage that complexity. I'm not against

> From an API perspective, I think that'd mean writing `trait MontgomeryPoint` as below, but without the methods. And then defining `struct AffineMontgomeryPoint<Base> { u: Base, v: Base }` and `impl<F> MontgomeryPoint<Base=F> for AffineMontgomeryPoint<F>` or something like that. Does that make sense?

The core method that the current trait needs is `from_base_coordinates`. That is necessary for safety: I want the coordinates struct to be guaranteed to represent a valid coordinate tuple, which requires having a validity-checked from-bare-coordinates constructor we can jump off to construct the coordinates struct.

So if we do split out a curve-only trait, we'd split this trait in the middle: have `trait MontgomeryCurve: Curve`, and then `trait AffineMontgomeryPoint: CurveAffine where Self::Curve: MontgomeryPoint`. Representing that trait bound in the type system is likely rather complex, but I'll have a go and see if it is workable.

Reply:
I think I understand. I'm all for keeping it simple. Do you think there's an easy way to have to two representations for an Edwards point? This has affine but curve25519-dalek needs XYZT

Reply:
`from_bare_coordinates` can convert to extended twisted Edwards coordinates by setting Z=1 and T=X*Y (and calling `is_valid` on the resulting point)

Reply:
> This is somewhat an artifact of the trait approach we've taken in the group crate: the Group trait denotes both a cryptographic group, and elements of that group. So there is currently no single place where a curve-only trait could live.

@str4d FWIW, in the `elliptic-curve` crate we have ZSTs which represent curve types and act as trait carriers (namely for traits that define the various associated element types)

Reply:
> Do you think there's an easy way to have to two representations for an Edwards point? This has affine but curve25519-dalek needs XYZT

@rozbb to clarify, it needs _to expose the coordinates of_ XYZT? If not and it just needs to use such points, that is already the case before this PR, and the idea behind the difference between the `Curve` and `CurveAffine` traits:
- `CurveAffine` represents specifically affine coordinates that are both closer to the encoding, and can be a type used for mixed addition or scalar mul.
- `Curve` represents the efficient type of a curve point for the given implementation; in the case of curve25519-dalek the XYZT representation.

If you want to expose the XYZT coordinates at all, then you can do that with concrete methods. If you want to expose them _generically_, then we'd need to discuss in what context this needs to be used:
- If the intent is to write generic code for the subset of efficient point impls that use XYZT coords, then extension traits on `Curve` that constrain to a given efficient model may be sufficient.
- If the intent is to obtain XYZT coords for given points, regardless of what coords they currently are in, then we'd need to define a trait or struct space that forms a state space of compatible coordinate models, along with an extension trait that allows converting the efficient `Curve` type from whatever its model is to the target model (which might be a no-op, or might not).

Reply:
> @str4d FWIW, in the `elliptic-curve` crate we have ZSTs which represent curve types and act as trait carriers (namely for traits that define the various associated element types)

Can you point me to these? I couldn't find them in the public API for the `elliptic-curve` crate.

Reply:
> @str4d FWIW, in the `elliptic-curve` crate we have ZSTs which represent curve types and act as trait carriers (namely for traits that define the various associated element types)

@tarcieri Can you point me to these? I couldn't find them in the public API for the `elliptic-curve` crate.

Reply:
This is the main one (and also one of the main integration points with the `group` crate, as it were): https://docs.rs/elliptic-curve/latest/elliptic_curve/trait.CurveArithmetic.html

Reply:
@str4d  you're right, we don't need to expose our specific representation. Any equivalent one should do, at least for a first draft